### PR TITLE
Proof of concept search methods for orders.

### PR DESCRIPTION
**Pull request by emdashcodes**
See #12677.

This PR is not complete. I'm posting this for feedback on how it works before going further, rather then a code review. There are still more todos, I want to cleanup the code a bit, and of course we should do the same for products (Should all CRUD classes have a search function?).

Essentially, the data store will have a `search` method that can be overridden. `wc_search_orders` will serve as the alias for the method.

`wc_search_orders` accepts a list of args/props and a return type. We should also support pagination but this PR doesn't do that yet. All fields passed that match a valid prop are searched as such, and anything else is treated as custom data.

The CPT implementation will pass through to WP_Query instead of trying to construct our own fragile queries.

The syntax for `wc_search_orders` is similar to WP_Query so it should be familiar to WP developers, but also easily hooked into so the fields can be passed to a different engine.

Here are two examples:

```
wc_search_orders( array(
    'customer_id' => array( 'value' => 1, 'compare' => '!=' ),
    'status' => 'cancelled',
    'date_created' => array( 'year' => '2017' )
) );
```

This will search for canceled orders, created in 2017, for all customers except customer `1`.

```
wc_search_orders( array( 'customer_id' => 1, 'status' => 'pending', 'my_custom_field' => 'test' ) );
```

This will search for all pending orders for customer `1`, with the custom meta field `test`.

### CODE REVIEW

File: includes/wc-order-functions.php [hunk 1]
```diff
@@ -616,15 +616,28 @@ function wc_order_fully_refunded( $order_id ) {
...
+ * Search orders by term.
  *
  * @since  2.6.0
  * @param  string $term Term to search.
  * @return array List of orders ID.
  */
 function wc_order_search( $term ) {
 	$data_store = WC_Data_Store::load( 'order' );
-	return $data_store->search_orders( str_replace( 'Order #', '', wc_clean( $term ) ) );
+	return $data_store->search_orders_by_term( str_replace( 'Order #', '', wc_clean( $term ) ) );
```
Comment (claudiosanches) [inline includes/wc-order-functions.php]: This should have support for translations. cc @mikejolley

**Comment by claudiosanches** [comment]
Not a comment specific for this Pull Request... But I think we could handle search inside `wc_get_orders()` so no need for two new functions doing it.

**Comment by justinshreve** [comment]
I don't like how return/pagination/orderby is mixed in with the fields for `wc_get_orders`, but I agree the logic is similar. `wc_get_orders` could call the search method so we don't duplicate code.

**Comment by mikejolley** [comment]
Interesting, yes it looks like this will do the trick.

> I don't like how return/pagination/orderby is mixed in with the fields for wc_get_orders, but I agree the logic is similar.

Agreed. I don't think we should mix them to the extent it's hard to follow each separate method (the 2 functions have different purposes) but if they can share some logic (e.g. the WP_Query generation, even if thats private) that could be good.

**Comment by mikejolley** [comment]
Do we need to worry about and vs or type searches, or would that be too complicated? I assume it's 'and' logic right now (match all search props).

**Comment by justinshreve** [comment]
> Do we need to worry about and vs or type searches, or would that be too complicated? I assume it's 'and' logic right now (match all search props).

Yeah we should support it, I don't think it would be too complicated. And yeah right now it's using 
'and' logic.

**Comment by justinshreve** [comment]
Thanks guys! I'm going to go ahead and close this and I'll reopen something later.
